# 文件统计：

In [99]:
import os
from tqdm import tqdm
path = './'
#os.chdir(path)
import re


filenames= os.listdir()
file_list = [ filename  for filename in filenames if filename.endswith('.txt') and filename.islower()]
print('total:',len(file_list))
file_list

total: 11


['aishell_test.txt',
 'chunks.txt',
 'aishell_dev.txt',
 'stcmd.txt',
 'aishell_train.txt',
 'aidatatang.txt',
 'prime.txt',
 'thchs_train.txt',
 'thchs_test.txt',
 'thchs_dev.txt',
 'trans.txt']

In [100]:
file_list[1:]

['chunks.txt',
 'aishell_dev.txt',
 'stcmd.txt',
 'aishell_train.txt',
 'aidatatang.txt',
 'prime.txt',
 'thchs_train.txt',
 'thchs_test.txt',
 'thchs_dev.txt',
 'trans.txt']

# 规范数据格式校验代码：

In [106]:
data_file = 'train.tsv'

pny_lst = []
han_lst = []
with open(data_file,'w+',encoding='utf-8') as final_file:
    
    for file in file_list[1:]:
        print(file)
        
        with open(file, 'r', encoding='utf8') as f:
            data = f.readlines() 
            print(len(data))
            for line in tqdm(data):
                wav_file, pny, han = line.split('\t')               
                
                pny = pny.split()
                han = han.strip()
                temp_pny = []
                temp_han = []
                flag = True
                for pny_item in pny:

                    if re.match('\d+', pny_item):
                        flag = False
                        break
                    else:
                        temp_pny.append(pny_item)

                for han_item in han:
                    if re.match('\d+', han_item):
                        flag = False
                        break
                    else:
                        temp_han.append(han_item)

                if len(temp_pny)!= len(temp_han):
#                     print(pny)
#                     print(len(temp_pny),'------',temp_pny)
#                     print(han)
#                     print(len(temp_han),'------',temp_han)
                    flag = False

                if flag:
                    final_file.write(wav_file+'<SEP>'+' '.join(temp_pny)+'<SEP>'+''.join(temp_han).strip()+'\n')


  0%|          | 0/14326 [00:00<?, ?it/s]

chunks.txt
119
aishell_dev.txt
14326


  0%|          | 0/102600 [00:00<?, ?it/s]

stcmd.txt
102600


  0%|          | 0/120098 [00:00<?, ?it/s]

aishell_train.txt
120098


100%|██████████| 120098/120098 [00:05<00:00, 22381.41it/s]


aidatatang.txt


  1%|▏         | 3250/237265 [00:00<00:07, 32492.98it/s]

237265


  0%|          | 0/50902 [00:00<?, ?it/s]

prime.txt
50902


 12%|█▏        | 1161/10000 [00:00<00:00, 11602.76it/s]

thchs_train.txt
10000


 43%|████▎     | 1067/2495 [00:00<00:00, 10653.88it/s]

thchs_test.txt
2495


100%|██████████| 893/893 [00:00<00:00, 8686.09it/s]


thchs_dev.txt
893
trans.txt


  0%|          | 2222/585273 [00:00<00:26, 22218.01it/s]

585273


100%|██████████| 585273/585273 [00:17<00:00, 33505.02it/s]


# 检测出问题数据：

file_list[4] 6  9

In [126]:
for i,item in enumerate(file_list):
    if i in [4,6,9]:
        print(item)

TRANS.txt
aidatatang.txt
cv.syllable.txt


### TRANS.txt:
1. UtteranceID	SpeakerID	Transcription
2. 14_3466_20170826171159.wav	14_3466	请语言播放小说
3. 14_3466_20170826171236.wav	14_3466	这里
4. 14_3466_20170826171323.wav	14_3466	全民唱吧

### cv.syllable.txt
1. A11_101 qi1 shi2 nian2 dai4 mo4 wo3 wai4 chu1 qiu2 xue2 mu3 qin1 ding1 ning2 wo3 chi1 fan4 yao4 xi4 jiao2 man4 yan4 xue2 xi2 yao4 shen1 zuan1 xi4 yan2 
2. A11_119 chen2 yun2 tong2 zhi4 tong2 shi2 yao1 qiu2 gan4 bu5 ren4 zhen1 xue2 xi2 ye4 wu4 jing1 tong1 ye4 wu4 xiang4 yi2 qie4 ye4 wu4 nei4 hang2 ren2 xue2 xi2 

### aidatatang.txt
1. aidatatang/G0002/session01/T0055G0002S0001.wav	yi3 hou4 ni3 shi4 nan2 hai2 zi	以后你是男孩子
2. aidatatang/G0002/session01/T0055G0002S0002.wav	lan2 zhou1 na3 you3 mai3 lu4 hu3 qi4 che1 de	兰州哪有买路虎汽车的
3. aidatatang/G0002/session01/T0055G0002S0004.wav	kan4 kan4 wo3 de ri4 cheng2 biao3	看看我的日程表

### aidatatang.txt 规范以后
237265  -->>  232759  数据减少1.89%

# 正则表达式

In [ ]:
import re
re.match(pattern, string, flags=0)

re.search(pattern, string, flags=0)  # re.search 扫描整个字符串并返回第一个成功的匹配

re.sub(pattern, repl, string, count=0, flags=0)# Python 的 re 模块提供了re.sub用于替换字符串中的匹配项。


re.compile 函数
compile 函数用于编译正则表达式，生成一个正则表达式（ Pattern ）对象，供 match() 和 search() 这两个函数使用。

语法格式为：

>>>import re
>>> pattern = re.compile(r'\d+')                    # 用于匹配至少一个数字
>>> m = pattern.match('one12twothree34four')        # 查找头部，没有匹配
>>> print m

 
pattern = re.compile(r'\d+')   # 查找数字
result1 = pattern.findall('runoob 123 google 456')
result2 = pattern.findall('run88oob123google456')

# 单个文件测试

In [107]:
data_file = 'test.tsv'

pny_lst = []
han_lst = []
with open(data_file,'w+',encoding='utf-8') as final_file:
    
        
    with open(file_list[0], 'r', encoding='utf8') as f:
        data = f.readlines() 
        print(len(data))
        for line in tqdm(data):
            wav_file, pny, han = line.split('\t')               

            pny = pny.split()
            han = han.strip()
            temp_pny = []
            temp_han = []
            flag = True
            for pny_item in pny:

                if re.match('\d+', pny_item):
                    flag = False
                    break
                else:
                    temp_pny.append(pny_item)

            for han_item in han:
                if re.match('\d+', han_item):
                    flag = False
                    break
                else:
                    temp_han.append(han_item)

            if len(temp_pny)!= len(temp_han):
                print(pny)
                print(len(temp_pny),'------',temp_pny)
                print(han)
                print(len(temp_han),'------',temp_han)
                flag = False

            if flag:
                final_file.write(wav_file+'<SEP>'+' '.join(temp_pny)+'<SEP>'+''.join(temp_han).strip()+'\n')

        



 25%|██▌       | 1811/7176 [00:00<00:00, 18109.59it/s]

7176


100%|██████████| 7176/7176 [00:00<00:00, 18562.03it/s]


# 读取文件

返回二维拼音 list

二维汉字list

In [16]:
def read_file(filename):
    """读取文件数据"""

    with open(filename,'r', encoding='utf-8') as f:
        lines = f.readlines()
        pny_list = []
        han_list = []
        for line in lines:
            pny_temp=[]
        
            wav_file, pny, han  = line.split('<SEP>')
            han = han.strip()
            for pny_item in pny.split(' '):
                pny_temp.append(pny_item)
                
            pny_list.append(pny_temp)
            han_list.append([item for item in han])

        han_list= [han for han  in han_list]
    return pny_list, han_list

In [46]:
pny_list,han_list = read_file('train.tsv')

In [47]:
pny_list[0]

['guang3',
 'zhou1',
 'shi4',
 'fang2',
 'di4',
 'chan3',
 'zhong1',
 'jie4',
 'xie2',
 'hui4',
 'fen1',
 'xi1']

# 构建词典

han_dict_w2id = defaultdict(int)
for index, item in enumerate(han_vocab):
    han_dict_w2id[item] = index

han_dict_id2w = {v: k for k, v in han_dict_w2id.items()}


pny_dict_w2id = defaultdict(int)
for index, item in enumerate(pny_vocab):
    pny_dict_w2id[item] = index

pny_dict_id2w = {v: k for k, v in pny_dict_w2id.items()}



In [48]:
from collections import defaultdict
import random

def mk_lm_pny_vocab(data):
    vocab = ['<PAD>']
    shuffle_vocab = []
    for line in tqdm(data):
        for pny in line:
            if pny not in shuffle_vocab:
                shuffle_vocab.append(pny)
    random.shuffle(shuffle_vocab)
    vocab.extend(shuffle_vocab)
    pny_dict_w2id = defaultdict(int)
    for index, item in enumerate(vocab):
        pny_dict_w2id[item] = index

    pny_dict_id2w = {v: k for k, v in pny_dict_w2id.items()}

    return pny_dict_w2id, pny_dict_id2w


def mk_lm_han_vocab(data):
    vocab = ['<PAD>']
    shuffle_vocab = []
    for line in tqdm(data):
        # line = ''.join(line.split(' '))
        for han in line:
            if han not in shuffle_vocab:
                shuffle_vocab.append(han)
                
    random.shuffle(shuffle_vocab)
    vocab.extend(shuffle_vocab)
    
    han_dict_w2id = defaultdict(int)
    for index, item in enumerate(vocab):
        han_dict_w2id[item] = index

    han_dict_id2w = {v: k for k, v in han_dict_w2id.items()}

    return han_dict_w2id, han_dict_id2w

In [49]:
pny_dict_w2id, pny_dict_id2w = mk_lm_pny_vocab(pny_list)
han_dict_w2id, han_dict_id2w = mk_lm_han_vocab(han_list)

100%|██████████| 14326/14326 [00:04<00:00, 3161.25it/s]


In [60]:
def process_file(filename, pny_dict_w2id, han_dict_w2id, seq_len = 100):
    """将文件转换为id表示"""
    pny_list, han_list = read_file(filename)
    pny_id_list = []
    han_id_list = []
    for i in range(5):
        print(pny_list[i])
        pny_id_list.append([pny_dict_w2id[x] for x in pny_list[i]]+ [0]*(seq_len-len(pny_list[i])))
        han_id_list.append([han_dict_w2id[x] for x in han_list[i]]+ [0]*(seq_len-len(pny_list[i])))
        print(pny_id_list[i])
    return pny_id_list, han_id_list


In [29]:
import random
a = [1,2,3]
b = [4,5,6]
random.shuffle(a)
random.shuffle(b)
a

[3, 2, 1]

In [62]:
import tensorflow as tf

a=[
    [1,2],
    [3,4]    
]

tf.one_hot(a,depth=3)

<tf.Tensor 'one_hot:0' shape=(2, 2, 3) dtype=float32>

wav_file, pny, han 

### regularize TRANS.txt with pny

In [105]:
from pypinyin import lazy_pinyin ,Style
import re

with open('trans.txt','w', encoding='utf-8') as file:
    with open( 'TRANS.txt', 'r', encoding='utf8') as f:
        data = f.readlines()[1:]
        for line in tqdm(data):
            wav_file,_ , han = line.split('\t') 
            pny = ' '.join(lazy_pinyin(han, style=Style.TONE3))

            file.write(wav_file+'\t'+pny.strip('\n')+'\t'+han)

100%|██████████| 585273/585273 [02:04<00:00, 4712.89it/s]


In [103]:
from pypinyin import lazy_pinyin,  Style
hans = '今天天气真不错啊'
a = lazy_pinyin(hans, style=Style.TONE3)
s = ' '.join(lazy_pinyin(a, style=Style.TONE3))
s

'jin1 tian1 tian1 qi4 zhen1 bu4 cuo4 a'